README.md
Shrimp
======

Shrimp is a library of user-defined functions for simplifying the programming in Hadoop and Pig.
</p>
</p>
### NVL:
NVL lets you replace null (returned as a blank) with a string in the results of a query. 
If expr1 is null, then NVL returns expr2. If expr1 is not null, then NVL returns expr1.

    DEFINE NVL com.ttech.shrimp.NVL();
    -- input:
    -- (1,,2)
    A = LOAD '/data/smp' USING PigStorage(',') as (f1:int, f2:chararray, f3:chararray);
    -- NVL(exp1, exp2)
    -- NVL2(string1, value_if_null )
    B = FOREACH A GENERATE NVL(f1, '-1') as f1, NVL(f2, '-1') as f2, NVL(f3, '-1') as f3;
    -- output:
    -- (1, -1, 2)
    dump B;


Source: http://docs.oracle.com/cd/B28359_01/server.111/b28286/functions110.htm

### NVL2:
NVL2 lets you determine the value returned by a query based on whether a specified expression is null or not null. 
If expr1 is not null, then NVL2 returns expr2. If expr1 is null, then NVL2 returns expr3.

    DEFINE NVL2 com.ttech.shrimp.NVL2();
    -- input:
    -- (1,,2)
    -- NVL2(string1, value_if_NOT_null, value_if_null )
    A = LOAD '/data/smp' USING PigStorage(',') as (f1:int, f2:chararray, f3:chararray);
    -- NVL2(exp1, exp2, exp3);
    -- NVL2(string1, value_if_NOT_null, value_if_null )
    B = FOREACH A GENERATE NVL2(f1, f1, '-1');
    -- output:
    -- (1)
    dump B;

Source: http://docs.oracle.com/cd/E11882_01/server.112/e26088/functions120.htm


### SQOOP_DATE_FORMAT:
To be able to load Oracle, date format must be "DD-MM-YYYY HH24:MI:SS".
SQOOP_DATE_FORMAT lets you to convert the format to be able to load to the Oracle.
Input format must be "YYYYMMDDHH24MISS"

    DEFINE SQOOP_DATE_FORMAT com.ttech.shrimp.SQOOP_DATE_FORMAT();
    -- input:
    -- (20130905225900)
    A = LOAD '/data/smp' USING PigStorage(',') as (f1:int, f2:chararray, f3:chararray);
    B = FOREACH A GENERATE SQOOP_DATE_FORMAT(f1);
    -- output:
    -- (2013-09-05 22:59:00)
    dump B;


## Working with the source code

Here are some common tasks when working with the source code.

### Build the JAR

	ant build-jar
	


CHANGELOG
0.1
------------
NVL, NVL2, SQOOP_DATE_FORMAT added


src/com/ttech/shrimp/NVL.java
package com.ttech.shrimp;

import java.io.IOException;
import org.apache.pig.EvalFunc;
import org.apache.pig.backend.executionengine.ExecException;
import org.apache.pig.data.DataType;
import org.apache.pig.data.Tuple;

/**
 * NVL lets you replace null (returned as a blank) with a string in the results of a query. 
 * If expr1 is null, then NVL returns expr2. If expr1 is not null, then NVL returns expr1.
 * <p>
 * Example:
 * 
 * <pre>
 * {@code
 *  DEFINE NVL com.ttech.shrimp.NVL();
 * 
 * -- input:
 * -- (1,,2)
 * A = LOAD '/data/smp' USING PigStorage(',') as (f1:int, f2:chararray, f3:chararray);
 * B = FOREACH A GENERATE NVL(f1, '-1') as f1, NVL(f2, '-1') as f2, NVL(f3, '-1') as f3;
 * 
 *   -- output:
 *   -- (1, -1, 2)
 * dump B;
 * }
 * }
 * </pre>
 * 
 * </p>
 * 
 * @author "Burak ISIKLI <burak.isikli@gmail.com>"
 */
public class NVL extends EvalFunc<String> {
	public String exec(Tuple input) throws IOException {
		try {
			if (input instanceof Tuple) {
				String result = (String) input.get(1);
				if (input.get(0) == null)
					return result;
				else {
					// String output = (String) input.get(0);
					// return output;

					if (input.getType(0) == DataType.INTEGER)
						return Integer.toString((Integer) input.get(0));
					else if (input.getType(0) == DataType.CHARARRAY)
						return (String) input.get(0);
					else if (input.getType(0) == DataType.LONG)
						return Long.toString((Long) input.get(0));
					else if (input.getType(0) == DataType.FLOAT)
						return Float.toString((Float) input.get(0));
					else if (input.getType(0) == DataType.DOUBLE)
						return Double.toString((Double) input.get(0));
					else
						throw new RuntimeException("Input type error");

				}

			} else
				throw new IOException("Expected input to be Tuple, but got "
						+ input.getClass().getName());
		} catch (ExecException e) {
			throw e;
		}
	}
}

src/com/ttech/shrimp/NVL2.java
package com.ttech.shrimp;

import java.io.IOException;
import org.apache.pig.EvalFunc;
import org.apache.pig.backend.executionengine.ExecException;
import org.apache.pig.data.Tuple;

/**
 * If the value is null, replace it with the null value input, otherwise replace
 * it with the other value input,
 * <p>
 * Example:
 * 
 * <pre>
 * {@code
 *  DEFINE NVL2 com.ttech.shrimp.NVL2();
 * 
 * -- input:
 * -- (1,,2)
 * -- NVL2(string1, value_if_NOT_null, value_if_null )
 * A = LOAD '/data/smp' USING PigStorage(',') as (f1:int, f2:chararray, f3:chararray);
 * B = FOREACH A GENERATE NVL2(f1, f1, '-1');
 * 
 *   -- output:
 *   -- (1)
 * dump B;
 * }
 * }
 * </pre>
 * 
 * </p>
 * 
 * @author "Burak ISIKLI <burak.isikli@gmail.com>"
 */
public class NVL2 extends EvalFunc<String> {
	public String exec(Tuple input) throws IOException {
		try {
			if (input instanceof Tuple) {
				if (input.get(0) != null)
					return (String) input.get(1);
				else
					return (String) input.get(2);
			} else
				throw new IOException("Expected input to be Tuple, but got "
						+ input.getClass().getName());
		} catch (ExecException e) {
			throw e;
		}
	}
}

src/com/ttech/shrimp/MINUS.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.ttech.shrimp;

import static java.lang.String.format;

import java.io.IOException;
import java.util.HashSet;
import java.util.Iterator;
import java.util.Set;

import org.apache.pig.EvalFunc;
import org.apache.pig.backend.executionengine.ExecException;
import org.apache.pig.data.BagFactory;
import org.apache.pig.data.DataBag;
import org.apache.pig.data.Tuple;

/**
 * SUBTRACT takes two bags as arguments and returns a new bag composed of tuples of first bag not in the second bag.<br>
 * If null, bag arguments are replaced by empty bags.
 * <p>
 * The implementation assumes that both bags being passed to this function will fit entirely into memory simultaneously.
 * </br>
 * If that is not the case the UDF will still function, but it will be <strong>very</strong> slow.
 */
public class MINUS extends EvalFunc<DataBag> {

    /**
     * Compares the two bag fields from input Tuple and returns a new bag composed of elements of first bag not in the second bag.
     * @param input a tuple with exactly two bag fields.
     * @throws IOException if there are not exactly two fields in a tuple or if they are not {@link DataBag}.
     */
    @Override
    public DataBag exec(Tuple input) throws IOException {
        if (input.size() != 2) {
            throw new ExecException("SUBTRACT expected two inputs but received " + input.size() + " inputs.");
        }
        DataBag bag1 = toDataBag(input.get(0));
        DataBag bag2 = toDataBag(input.get(1));
        return subtract(bag1, bag2);
    }

    private static String classNameOf(Object o) {
        return o == null ? "null" : o.getClass().getSimpleName();
    }

    private static DataBag toDataBag(Object o) throws ExecException {
        if (o == null) {
            return BagFactory.getInstance().newDefaultBag();
        }
        if (o instanceof DataBag) {
            return (DataBag) o;
        }
        throw new ExecException(format("Expecting input to be DataBag only but was '%s'", classNameOf(o)));
    }

    private static DataBag subtract(DataBag bag1, DataBag bag2) {
        DataBag subtractBag2FromBag1 = BagFactory.getInstance().newDefaultBag();
        // convert each bag to Set, this does make the assumption that the sets will fit in memory.
        Set<Tuple> set1 = toSet(bag1);
        // remove elements of bag2 from set1
        Iterator<Tuple> bag2Iterator = bag2.iterator();
        while (bag2Iterator.hasNext()) {
            set1.remove(bag2Iterator.next());
        }
        // set1 now contains all elements of bag1 not in bag2 => we can build the resulting DataBag.
        for (Tuple tuple : set1) {
            subtractBag2FromBag1.add(tuple);
        }
        return subtractBag2FromBag1;
    }

    private static Set<Tuple> toSet(DataBag bag) {
        Set<Tuple> set = new HashSet<Tuple>();
        Iterator<Tuple> iterator = bag.iterator();
        while (iterator.hasNext()) {
            set.add(iterator.next());
        }
        return set;
    }

}


src/com/ttech/shrimp/SQOOP_DATE_FORMAT.java
package com.ttech.shrimp;

import java.io.IOException;

import org.apache.pig.EvalFunc;
import org.apache.pig.backend.executionengine.ExecException;
import org.apache.pig.data.DataType;
import org.apache.pig.data.Tuple;

/**
 * Formating the chararray for Sqoop Loading
 * <p>
 * Example:
 * 
 * <pre>
 * {@code
 *  DEFINE SQOOP_DATE_FORMAT com.ttech.shrimp.SQOOP_DATE_FORMAT();
 * 
 * -- input:
 * -- (20130905225900)
 * -- NVL2(string1, value_if_NOT_null, value_if_null )
 * A = LOAD '/data/smp' USING PigStorage(',') as (f1:int, f2:chararray, f3:chararray);
 * B = FOREACH A GENERATE SQOOP_DATE_FORMAT(f1);
 * 
 *   -- output:
 *   -- (2013-09-05 22:59:00)
 * dump B;
 * }
 * }
 * </pre>
 * 
 * </p>
 * 
 * @author "Burak ISIKLI <burak.isikli@gmail.com>"
 */
public class SQOOP_DATE_FORMAT extends EvalFunc<String> {

	public String exec(Tuple input) throws IOException {
		try {
			if (input instanceof Tuple) {
				if (input == null || input.size() < 1 || input.get(0) == null) {
					return null;
				}
				if (input.getType(0) != DataType.CHARARRAY)
					throw new ExecException(
							"Expected input's type to be chararray but got "
									+ input.getType(0));
				String strIn = DataType.toString(input.get(0));
				if (strIn.length() != 14)
					throw new ExecException(
							"Expected input's length to be 14 but got "
									+ strIn.length());
				String output = strIn.substring(0, 4) + '-'
						+ strIn.substring(4, 6) + '-' + strIn.substring(6, 8)
						+ ' ' + strIn.substring(8, 10) + ':'
						+ strIn.substring(10, 12) + ':'
						+ strIn.substring(12, 14);
				return output;
			} else
				throw new ExecException("Expected input to be Tuple, but got "
						+ input.getClass().getName());

		} catch (ExecException e) {
			throw e;
		}
	}
}